lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,211 @@
+    auto indexSecretness = indexSecretnessLattice
+                               ? indexSecretnessLattice->getValue()
+                               : Secretness();
+
+    // If lattice is set to unknown,
+    // apply transformation anyway but emit a warning
+    if (!indexSecretness.isInitialized()) {
+      extractOp->emitWarning()
+          << "Secretness for tensor.extract index is unknown. "
+             "Conservatively, the transformation will be applied:";
+    }
+
+    // If index is known to be public, no transformation is needed
+    if (indexSecretness.isInitialized() && !indexSecretness.getSecretness())
+      return failure();
+
+    ImplicitLocOpBuilder builder(extractOp->getLoc(), rewriter);
+
+    // Create index 0
+    auto zero = builder.create<arith::ConstantIndexOp>(0);
+    // Set secretness for index 0
+    setValueToSecretness(solver, zero, Secretness(false));
+
+    // Extract tensor value at index 0
+    SmallVector<Value> i = {zero};
+    auto initialValue =
+        builder.create<tensor::ExtractOp>(extractOp.getTensor(), i);
+    // Set secretness for initialValue
+    auto tensorSecretness =
+        solver->getOrCreateState<SecretnessLattice>(extractOp.getTensor())
+            ->getValue();
+    setValueToSecretness(solver, initialValue, tensorSecretness);
+
+    int size = extractOp.getTensor().getType().getShape().front();
+
+    SmallVector<Value> iterArgs = {initialValue};
+    auto forOp = builder.create<affine::AffineForOp>(0, size, 1, iterArgs);
+    // Set secretness for induction variable
+    setValueToSecretness(solver, forOp.getInductionVar(), Secretness(false));
+
+    builder.setInsertionPointToStart(forOp.getBody());
+
+    // Check if the current index is equal to the secret index
+    auto cond = builder.create<arith::CmpIOp>(arith::CmpIPredicate::eq,
+                                              forOp.getInductionVar(), index);
+    // Set secretness for cond
+    for (auto result : cond->getResults()) {
+      setValueToSecretness(solver, result, indexSecretness);
+    }
+
+    // Extract value at current index
+    auto newExtractOp = builder.create<tensor::ExtractOp>(
+        extractOp.getTensor(), forOp.getInductionVar());
+    // Set secretness for newExtractOp
+    for (auto result : newExtractOp->getResults()) {
+      setValueToSecretness(solver, result, tensorSecretness);
+    }
+
+    auto ifOp = builder.create<scf::IfOp>(
+        cond,
+        [&](OpBuilder &b, Location loc) {
+          // Yield value extracted at index
+          b.create<scf::YieldOp>(loc, newExtractOp.getResult());
+        },
+        [&](OpBuilder &b, Location loc) {
+          // Yield previous value
+          b.create<scf::YieldOp>(loc, forOp.getRegionIterArgs().front());
+        });
+    // Set secretness for ifOp results: Combine indexSecretness with
+    // tensorSecretness
+    auto combinedSecretness =
+        Secretness::join(indexSecretness, tensorSecretness);
```

Comment:
I don't think `join` is correct here. If one of the two is unknown/uninitialized and the other is "false", we want the combinedSecretness to also be unknown/uninitialized.
Effectively, you want the same thing as https://github.com/google/heir/blob/028e48f451a19d46a2dd18180b8c5fb51ffbe92c/lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp#L42-L68 
So my recommendation would be to factor that logic out into a separate function `combine` in the `Secretness` class that takes a range or SmallVector of `Secretness`. You can then use it both in `visitOperation` and in the Extract/Insert Conversion.


To go from the `ArrayRef<*SecretnessLattice> operands` you have in `visitOperation` to the right input for that method , you can use `llvm::to_vector<2>(llvm::map_range(operands, [](SecretnessLattice *v) -> Secretness(v->getValue())))` (a quick search in the repo will show plenty of examples, mine probably won't compile as-is).